From: Matthias Clasen <matthias.clasen@gmail.com>
Date: Thu, 28 May 2020 18:34:24 -0400
Subject: screenshot: Fix handling of cancellation (#312)

We were not properly passing a canceled PickColor
call back to the frontend.

Co-authored-by: Matthias Clasen <mclasen@redhat.com>
Origin: upstream, 1.7.2, commit:2bf4b92f8f818be8565ff3ed32de0d3c0bfccdfb
---
 src/screenshot.c | 30 +++++++++++++++++++-----------
 1 file changed, 19 insertions(+), 11 deletions(-)

diff --git a/src/screenshot.c b/src/screenshot.c
index 3ce1d89..0346207 100644
--- a/src/screenshot.c
+++ b/src/screenshot.c
@@ -228,24 +228,32 @@ color_picked (GObject *object,
   ScreenshotDialogHandle *handle = data;
   g_autoptr(GError) error = NULL;
   g_autoptr(GVariant) result = NULL;
-  
+
+  handle->response = 0;
+  handle->red = handle->green = handle->blue = 0;
+
   if (!org_gnome_shell_screenshot_call_pick_color_finish (shell, &result, res, &error))
     {
-      g_warning ("PickColor failed: %s", error->message);
-      return;
+      if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_CANCELLED))
+        {
+          g_warning ("PickColor cancelled");
+          handle->response = 1;
+        }
+      else
+        {
+          g_warning ("PickColor failed: %s", error->message);
+          handle->response = 2;
+        }
     }
-
-  if (!g_variant_lookup (result, "color", "(ddd)",
-                         &handle->red,
-                         &handle->green,
-                         &handle->blue))
+  else if (!g_variant_lookup (result, "color", "(ddd)",
+                              &handle->red,
+                              &handle->green,
+                              &handle->blue))
     {
       g_warning ("PickColor didn't return a color");
-      return;
+      handle->response = 2;
     }
 
-  handle->response = 0;
-
   send_response (handle);
 }
 
